### PR TITLE
sched: Unify the main thread and pthread behaivour

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
## Summary
1.Make pthread_cleanup_[push|pop] callable from the main thread
2.Robust mutex also support in main thread now

## Testing
nxstyle error can be ignored(mixed case and file rename):
```
include/stdlib.h:166:19: error: Mixed case identifier found
sched/task/task_atexit.c: fatal: Failed to open
sched/task/task_onexit.c: fatal: Failed to open
```

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@patacongo could you give some feedback? so other committer can merge the code.

**Comment by xiaoxiang781216** [comment]
Does anyone can help review the change?

File: sched/task/task_cancelpt.c [hunk 1]
```diff
@@ -339,51 +344,89 @@ void nxnotify_cancellation(FAR struct tcb_s *tcb)
...
+       *  in the target thread until the thread changes the cancellability.
+       *  When cancellability is deferred, all cancels are held pending in
+       *  the target thread until the thread changes the cancellability,
+       *  calls a function which is a cancellation point or calls
+       *  pthread_testcancel(), thus creating a cancellation point.  When
+       *  cancellability is asynchronous, all cancels are acted upon
+       *  immediately, interrupting the thread with its processing."
+       *
+       * REVISIT:  Does this rule apply to equally to both SIGKILL and
+       * SIGINT?
```
Comment (yamt) [inline sched/task/task_cancelpt.c:354]: i don't understand this REVISIT comment. how signals are related here?

File: sched/pthread/pthread_cancel.c [hunk 2]
```diff
@@ -85,6 +76,15 @@ int pthread_cancel(pthread_t thread)
       pthread_exit(PTHREAD_CANCELED);
     }
```
Comment (yamt) [inline sched/pthread/pthread_cancel.c:73]: does this mean to make pthread_cancel a cancellation point?

File: sched/task/task.h [hunk 3]
```diff
@@ -48,6 +48,13 @@
 
 #include <nuttx/sched.h>
 
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define EXIT_NORMAL                0
+#define EXIT_BYPASS                1
```
Comment (yamt) [inline sched/task/task.h]: can you document what this bypasses?

File: sched/task/task_delete.c [hunk 4]
```diff
@@ -143,6 +144,15 @@ int task_delete(pid_t pid)
       goto errout;
     }
 
+#ifdef HAVE_GROUP_MEMBERS
+  /* Kill all of the children of the task.  This will not kill the currently
+   * running task/pthread (this_task).  It will kill the main thread of the
+   * task group if this_task is a pthread.
+   */
+
+  group_kill_children(rtcb);
```
Comment (yamt) [inline sched/task/task_delete.c:153]: why is this necessary?
why is this rtcb, not dtcb?

File: sched/pthread/pthread_cleanup.c [hunk 5]
```diff
@@ -173,8 +165,7 @@ void pthread_cleanup_push(pthread_cleanup_t routine, FAR void *arg)
    */
 
   sched_lock();
-  if ((tcb->cmn.flags & TCB_FLAG_TTYPE_MASK) == TCB_FLAG_TTYPE_PTHREAD &&
-      tcb->tos < CONFIG_PTHREAD_CLEANUP_STACKSIZE)
+  if (tcb->tos < CONFIG_PTHREAD_CLEANUP_STACKSIZE)
```
Comment (btashton) [inline sched/pthread/pthread_cleanup.c:168]: Why are we not bailing out here when called with a non pthread type?

Comment (xiaoxiang781216) [reply sched/pthread/pthread_cleanup.c:168]: because main/kernel thread type is TCB_FLAG_TTYPE_TASK/TCB_FLAG_TTYPE_KERNEL, but both types should be able to call pthread_cleanup_push/pthread_cleanup_pop. The major target of whole patch set is ensure the main/kernel thread can call all most pthread API like pthread.

File: sched/task/task_delete.c (hunk 4 above)
Comment (xiaoxiang781216) [reply sched/task/task_delete.c:153]: > why is this necessary?

I think that It's dangerous to terminate the main thread but leave other threads in the task keep running. And other similar function also do the same thing(e.g. task_restart, nxsig_abnormal_termination, exit...).

> why is this rtcb, not dtcb?
good catch, I will fix it.

**Comment by xiaoxiang781216** [comment]
> this PR is complex.

Yes, it's better to review the patch one by one in commit tab.

> can you consider to separate quick_exit into another PR?

It's too difficult because quick_exit depend the previous patch which clean up the exit logic. If need, I  can remove quick_exit patch from the PR and send a new PR once this PR get merged.

File: sched/task/task_cancelpt.c (hunk 1 above)
Comment (xiaoxiang781216) [reply sched/task/task_cancelpt.c:354]: The comment is copied from sig_default, I will remove it.

File: sched/task/task.h (hunk 3 above)
Comment (xiaoxiang781216) [reply sched/task/task.h]: Ok.

File: sched/pthread/pthread_cancel.c (hunk 2 above)
Comment (xiaoxiang781216) [reply sched/pthread/pthread_cancel.c:73]: The behaviour is same as before, the change is just an optimizition. If you look at nxnotify_cancellation, the main job is call nxsem_wait_irq/nxsig_wait_irq/nxmq_wait_irq but the code unnecessary if some thread call pthread_cancel on self because the thread is impossible in the wait state.

File: sched/task/task_delete.c (hunk 4 above)
Comment (yamt) [reply sched/task/task_delete.c]: why is it dangerous? it's something i believe the most of pthread implementations allow.

File: sched/pthread/pthread_cancel.c (hunk 2 above)
Comment (yamt) [reply sched/pthread/pthread_cancel.c:73]: i don't think pthread_cancel(self) is supposed to terminate the thread in the case of deferred cancellation.
after your changes, it seems to do.

**Comment by davids5** [comment]
@xiaoxiang781216 why is this needed? Why did you take the approach of removing the union? Where are the test cases with all the permutations of the CONFIG_* that control the changes?

File: sched/pthread/pthread_cleanup.c (hunk 5 above)
Comment (xiaoxiang781216) [reply sched/pthread/pthread_cleanup.c:168]: > To me it is a totally unacceptable partitioning of user space and kernel interfaces. I am very opposed to supported this. It is wrong. It is a serious architectural mistake. Let's not go that way... it is very sickening design concept. Let's keep logic in place to prohibit the use of any pthread APIs from kernel threads.

Ok, we can make the change like this:
```
DEBUGASSERT((tcb->flags & TCB_FLAG_TTYPE_MASK) !=
                TCB_FLAG_TTYPE_KTHREAD);
```
or
```
if ((rtcb->flags & TCB_FLAG_TTYPE_MASK) != TCB_FLAG_TTYPE_KTHREAD)
{
...
}
```
So the kernel thread will fail or no-op for pthread API, but main/pthread can work correctly.

> Please. Don't shit on the OS! We are entrusted you to maintain good functional partitioning and not to trash all of the interfaces.
> 
> I am super opposed to the concept. I am very disappointed in what you are doing to the OS. Very disappointed.

> My long term goals include:
> 
> 1. To separate kernel threads from all groups.  To eliminate streams and file descriptors from strings,
> 2. Move all pthread logic possible into user space as will all other Unix system.
> 
> You are ruining that.

This change doesn't ruining your goal, but a path to your final target:
1.Unify the main and thread behaviour:
   a.main thread can call pthread_cleanup_[push|pop]
   b.robust mutex work with main thread too
   c.pthread_exit in main shouldn't terminate the whole process
   b.cancellation point in main shouldn't terminate the whole process
2.The next step is reorg tcb_s maybe like this:
```
            tcb_s
              |
      +-------+--------+
      |                |
tcb_kernel_s     tcb_user_s
                    |
            +-------+--------+
            |                |
        tcb_task_s     tcb_pthread_s
``` 
3.Finally merge tls_info_s into tcb_user_s and place userspace tcb at the beginning of stack
You can see the first step help this path because it make main thread same as pthread which is always required.

File: include/nuttx/sched.h [hunk 6]
```diff
@@ -714,6 +714,23 @@ struct tcb_s
...
+
+#ifdef CONFIG_PTHREAD_CLEANUP
+  /* tos   - The index to the next available entry at the top of the stack.
+   * stack - The pre-allocated clean-up stack memory.
+   */
+
+  uint8_t tos;
+  struct pthread_cleanup_s stack[CONFIG_PTHREAD_CLEANUP_STACKSIZE];
+#endif
+
```
Comment (patacongo) [inline include/nuttx/sched.h:733]: Wouldn't it be better to move these into TLS?  They should only be used by user space applications and, hence, should be accessed in the same way that the errno variable is.

Comment (xiaoxiang781216) [reply include/nuttx/sched.h:733]: @patacongo I will provide a patch to move them to TLS when I finish upgrade libcxx to the latest version.
